Implement object lock manager for equivocation prevention

lxfind opened the pull request:
# Object Locks for Gas Pool Transactions

## Overview
Introduces an object locking mechanism to prevent transaction equivocation in the gas pool service by ensuring concurrent transactions cannot modify the same address-owned objects simultaneously. We fail immediately on lock acquisition failure (rather than waiting/retrying) since the other transaction will mutate the locked object and advance its version, meaning this transaction would never succeed to execute.

## Key Components
- `ObjectLockManager`: Manages locks for address-owned objects during transaction execution using RAII-style locking
- Integrated with `GasPool` to acquire locks before transaction execution
- Selectively locks only address-owned objects, excluding gas objects, immutable objects, and shared objects
- Uses caching and batching for efficient ownership checks

## Testing
Comprehensive test suite covering basic lock acquisition/release, concurrent scenarios, and edge cases.

amnn commented on src/object_locks.rs:
```diff
@@ -0,0 +1,448 @@
+    address_owned_cache: SegmentedCache<ObjectID, bool>,
+    /// A cache that tracks the objects that are currently locked due to
+    /// active execution of a transaction.
+    locked_owned_objects: Arc<RwLock<HashSet<ObjectID>>>,
```
I think this could just be a `Mutex` -- there's no occasion where you want to just read the state here along with other readers -- you would only be reading it in order to later write to it.

amnn commented on src/object_locks.rs:
```diff
@@ -0,0 +1,448 @@
+/// This can help avoid equivocation of transactions that concurrently modify
+/// the same address-owned objects.
+pub struct ObjectLockManager {
+    /// A cache that tracks whether an object is address-owned.
+    /// Using SegmentedCache for better concurrent performance.
+    address_owned_cache: SegmentedCache<ObjectID, bool>,
```
nit: could you document what the `bool` here is? (immutable vs address-owned, right?)

amnn commented on src/object_locks.rs:
```diff
@@ -0,0 +1,448 @@
+
+        for obj in objects {
+            match self.address_owned_cache.get(&obj) {
+                Some(true) => mutable_objects.push(obj),
+                Some(false) => {} // Ignore known immutable objects
+                None => objects_to_query.push(obj),
```
Hmm, is this always safe?

- Owned objects can become frozen later, so we may accidentally add an object into `mutable_objects` and try and acquire a lock on it and fail unnecessarily when in fact it's frozen and it's fine to access from multiple transactions concurrently.
- I got the impression that with the introduction of consensus-managed objects, an object could transition between consensus-managed and address-owned.

AFAICT the only thing you can cache by just their ID is objects that are known to be immutable -- everything else you may need to check every time.

You may be able to improve performance further by populating this cache based on the effects from transactions that were executed by the gas pool, and keying it by both the object ID and version. This technique also has the nice property that we also get to cache the ownership of shared/consensus-managed objects. As long as all the transactions that touch it go through the gas pool, we'll get a good cache hit rate.

lxfind replied:
Good catch.

> As long as all the transactions that touch it go through the gas pool, we'll get a good cache hit rate.

I don't know if we could count on this though.
Worst case we will just make more object queries I guess. Let me take a look.
Thanks for review!

lxfind replied:
Probably not even necessary. In the update PR I make sure it updates the cache post execution, which should handle all cases eventually

amnn replied:
Here's the scenario I'm worried about (which motivates using versions in cache keys):

- Sponsored TX on Owned object O (ownership unchanged): `O` is cached as address owned.
- External transaction on the same owned object, freezes it.
- Two sponsored transactions come in touching `O` at the same version. Technically fine because it's frozen now, but the gas pool will complain because its cache tells it that its owned and locked to one of those transactions.

lxfind replied:
We would however need to be careful not bloating the cache (we need to be able to remove entires for old versions)

amnn replied:
Indeed, I was imagining the cache would be LRU, and when a transaction settles we would remove global cache entries from it that correspond to the transaction's inputs (and populate the cache with new entries from its outputs)

amnn commented on src/object_locks.rs:
```diff
@@ -0,0 +1,448 @@
+        assert_eq!(locks2.locked_objects[0], obj_id);
+    }
+
+    #[tokio::test]
```
You should have a test for an object that transitions from owned to immutable, and then is accessed concurrently.

amnn reviewed (commented):
I think there might be some edge cases related to the owner cache, PTAL. I suggested an alternative.

lxfind commented:
Addressed feedback.
The cache still caches whether a given object is address-owned. Then we provide opportunities to update the cache to deal with any potential inconsistencies post execution.

amnn commented on src/object_locks.rs:
```diff
@@ -0,0 +1,530 @@
+    /// By removing them from the cache, we will re-query the owner of the object and update
+    /// the cache next time we see the object. This also helps reduce the cache size
+    /// by removing deleted or wrapped objects.
+    pub fn update_cache_post_execution(
```
I'm not sure this is enough, is it still correct in case someone modifies ownership through a transaction outside of the gas pool? That's why I proposed both updating the cache post-effects and keying the cache by ID and version -- the version in the cache key helps prevent us using a stale cache entry in that scenario.

lxfind replied:
For example, if someone freezes an object outside of the pool, the first time the pool sees this object used after that, it will lock it thinking it is an owned object, but then as soon as the first transaction executes, it will be able to update the cache, because it now sees the input not mutated.

lxfind commented:
Ok added version tracking in the cache.

amnn reviewed (approved):
LGTM! Thanks @lxfind
